go/hyperkit.go

```diff
@@ -281,43 +264,29 @@ func (h *HyperKit) execute(cmdline string) error {
 		}
 	}
 
-	// Create files
-	if h.StateDir != "" {
-		err = os.MkdirAll(h.StateDir, 0755)
-		if err != nil {
-			return err
+	// Create directories.
+	for _, dir := range []string{h.VSockDir, h.StateDir} {
+		if dir != "" {
+			if err := os.MkdirAll(dir, 0755); err != nil {
+				return fmt.Errorf("Cannot create directory %q: %v", dir, err)
+			}
 		}
 	}
 
-	for idx, config := range h.Disks {
-		if config.Path == "" {
+	for idx, disk := range h.Disks {
+		if disk.GetPath() == "" {
 			if h.StateDir == "" {
 				return fmt.Errorf("Unable to create disk image when neither path nor state dir is set")
 			}
-			if config.Size <= 0 {
+			if disk.GetSize() <= 0 {
 				return fmt.Errorf("Unable to create disk image when size is 0 or not set")
 			}
-			config.Path = filepath.Clean(filepath.Join(h.StateDir, fmt.Sprintf("disk%02d.img", idx)))
-			h.Disks[idx] = config
+			disk.SetPath(filepath.Clean(filepath.Join(h.StateDir, fmt.Sprintf("disk%02d.img", idx))))
+			h.Disks[idx] = disk
 		}
-
-		// assume the path could be a file:// uri with query params
-		// use the actual path from the parsed result
-		u, err := url.Parse(config.Path)
-		if err != nil {
```

Comment:
This does not seem to have moved into the `RawDisk` handler. Is that an error introduced by rebasing over #168 or something deliberate? I think this is going to have an effect on the `linuxkit` tool's use of this interface, which currently allows the user to simply pass in the path (which might be a `file:///path` thing AIUI).

I don't think we want to teach `linuxkit` to parse those URLs in order to construct a `QcowDisk` object, or add a lot of options to the CLI so users can specify all the options independently in order to construct it, so I think we need to continue to support `RawDisk` where the `.Path` is structured as a URL still. I suspect (but I'm not sure) that this is just a case of replacing the `url.Parse` in the `Exists()` function.

Reply:
I had no realised that it was considered a _feature_ to have this.  The current code does not even checks that it's not say `https://` which is used.

What exactly should be the interface?

Reply:
`linuxkit run hyperkit -disk «foo»` needs to keep working for both raw and qcow disks.

This code was, AIUI, just checking that the path part was valid. Whether or not the schema was valid is not checked until hyperkit itself tries to use it, when it will probably reject `https` (unless @djs55's qcow library is cleverer than I imagined). This was only added earlier in the week in #168.

Reply:
I am really reluctant to trying to force this URI into the Path.  The whole point of the interface was precisely to make this syntactic difference go away.

So am I correctly understanding that linuxkit's interface for disk expect either `foo.raw` or `file:///foo.qcow`?  Shouldn't it rather rely on conventions on the file extension?

I suppose the best option would be something like `MakeDisk` that hides the instantiation of RawDisk/QcowDisk, and I certainly agree it would be a nice addition, but it's its input which is still unclear to me.  If I could see a few concrete examples of `«foo»` in `linuxkit run hyperkit -disk «foo»`, I'll do that.  How you want to specify options for raw and for qcow would help me.

Reply:
The intention of https://github.com/moby/hyperkit/pull/168 was to allow file:// uris to be passed to hyperkit. This was an expectation of hyperkit for qcow files

Reply:
> So am I correctly understanding that linuxkit's interface for disk expect either foo.raw or file:///foo.qcow? Shouldn't it rather rely on conventions on the file extension?

It expects the same things the hyperkit CLI does I think since apart from deconstructing it to check the path element exists it just passes it through in (before you PR) `fmt.Sprintf("%d:0,%s,%s", nextSlot, driver, p.Path)`.

So assuming I've understood the CLI correctly (cc @djs55), yes, it expects `foo.raw` or `file:///foo.qcow` or `file://foo.qcow2?someoption=yes`.